### PR TITLE
Feat tapd

### LINKED ISSUE #560: Add TAPD as our data source
## Description
TAPD is used by many of our customer, it should be supported.

## Describe the solution you'd like
Develop a plugin to import data from TAPD

## Has the Feature been Requested Before?
Yes

### PR DESCRIPTION
**Pull request by warren830**
# Summary

tapd testconnection, source, workspace

# Related Issue
close #560 

### Screenshots
sprint issues
image
users
image
worklog
image
board_issues
image
board_sprints
image
boards
image
changelogs
image
issue_assignee_history
image
issue_commits
image
issue_sprints_history
image
issue_status_history
image
issues
image
sprint_issues
image
sprints
image

### Other Information
Any other information that is important to this PR.

### CODE REVIEW

File: plugins/tapd/tasks/api_client.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	// create synchronize api client so we can calculate api rate limit dynamically
+	headers := map[string]string{
+		"Authorization": fmt.Sprintf("Basic %v", auth),
+	}
+	apiClient, err := helper.NewApiClient(source.Endpoint, headers, 0, source.Proxy, taskCtx.GetContext())
+	if err != nil {
+		return nil, err
+	}
+	apiClient.SetAfterFunction(func(res *http.Response) error {
+		if res.StatusCode == http.StatusUnprocessableEntity {
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/tapd/tasks/api_client.go:29]: Yes, I just double checked

File: plugins/tapd/models/migrationscripts/archived/tapd_source.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	Name                       string `gorm:"type:varchar(100);uniqueIndex" json:"name" validate:"required"`
+	Endpoint                   string `gorm:"type:varchar(255)"`
+	BasicAuthEncoded           string `gorm:"type:varchar(255)"`
+	EpicKeyField               string `gorm:"type:varchar(50);" json:"epicKeyField"`
+	StoryPointField            string `gorm:"type:varchar(50);" json:"storyPointField"`
+	RemotelinkCommitShaPattern string `gorm:"type:varchar(255);comment='golang regexp, the first group will be recognized as commit sha, ref https://github.com/google/re2/wiki/Syntax'" json:"remotelinkCommitShaPattern"`
+	Proxy                      string `gorm:"type:varchar(255)"`
+	RateLimit                  int    `comment:"api request rate limt per second"`
+	//CompanyId                  uint64 `json:"companyId" validate:"required"`
+	WorkspaceId uint64 `json:"workspaceId" validate:"required"`
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/tapd/models/migrationscripts/archived/tapd_source.go]: Already fixed

File: plugins/core/iso8601time.go [hunk 3]
```diff
@@ -150,3 +161,17 @@ func DecodeMapStruct(input map[string]interface{}, result interface{}) error {
 	}
 	return err
 }
+
+func ConvertStringToTimePtr(timeString string) (*time.Time, error) {
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/core/iso8601time.go:169]: I think I met one before, but now we use ISO8601, I can delete this.

File: plugins/tapd/api/tapd_connection.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+			"Authorization": fmt.Sprintf("Basic %s", token),
+		},
+		3*time.Second,
+		params.Proxy,
+		nil,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("verify token failed for %s %w", token, err)
+	}
+	res, err := apiClient.Get("/quickstart/testauth", nil, nil)
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/tapd/api/tapd_connection.go:61]: It returned 401, I fixed it by adding code as  below:
	res, err := apiClient.Get("/quickstart/testauth", nil, nil)
	if err != nil {
		return nil, err
	}
	if res.StatusCode == http.StatusUnauthorized {
		return nil, fmt.Errorf("verify token failed for %s", token)
	}
	if res.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status code: %d", res.StatusCode)
	}

File: plugins/tapd/api/tapd_sources.go [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+}
+
+// GET /plugins/tapd/sources/:sourceId/epics
+type EpicResponse struct {
+	Id    int
+	Title string
+	Value string
+}
+
+func GetEpicsBySourceId(input *core.ApiResourceInput) (*core.ApiResourceOutput, error) {
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/tapd/api/tapd_sources.go:257]: fixed

File: plugins/tapd/tasks/api_client.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+			return fmt.Errorf("authentication failed, please check your Basic Auth Token")
+		}
+		return nil
+	})
+
+	// create rate limit calculator
+	rateLimiter := &helper.ApiRateLimitCalculator{
+		UserRateLimitPerHour: source.RateLimit,
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/tapd/tasks/api_client.go:37]: I added this when save connection. And add ratelimit to the table.source

File: plugins/tapd/models/tapd_source.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import (
+	"github.com/merico-dev/lake/models/common"
+)
+
+type TapdSource struct {
+	common.Model
+	Name                       string `gorm:"type:varchar(100);uniqueIndex" json:"name" validate:"required"`
+	Endpoint                   string `gorm:"type:varchar(255)"`
+	BasicAuthEncoded           string `gorm:"type:varchar(255)"`
+	EpicKeyField               string `gorm:"type:varchar(50);" json:"epicKeyField"`
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/tapd/models/tapd_source.go:12]: fixed

File: plugins/tapd/models/migrationscripts/archived/tapd_source.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"github.com/merico-dev/lake/models/common"
+)
+
+type TapdSource struct {
+	common.Model
+	Name                       string `gorm:"type:varchar(100);uniqueIndex" json:"name" validate:"required"`
+	Endpoint                   string `gorm:"type:varchar(255)"`
+	BasicAuthEncoded           string `gorm:"type:varchar(255)"`
+	EpicKeyField               string `gorm:"type:varchar(50);" json:"epicKeyField"`
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/tapd/models/migrationscripts/archived/tapd_source.go]: fixed

File: plugins/tapd/tasks/api_client.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+		rateLimiter,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return asyncApiClient, nil
+}
+
+func NewTapdApiPageClient(taskCtx core.TaskContext, source *models.TapdSource) (*helper.ApiClient, error) {
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/tapd/tasks/api_client.go:51]: Cause tapd get total pages by another api

[the remark this answers is not in the document]
Comment (warren830) [reply plugins/tapd/tasks/api_client.go:51]: fixed

File: plugins/tapd/api/tapd_sources.go [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+	if encKey == "" {
+		// Randomly generate a bunch of encryption keys and set them to config
+		encKey = core.RandomEncKey()
+		v.Set(core.EncodeKeyEnvStr, encKey)
+		err := v.WriteConfig()
+		if err != nil {
+			return err
+		}
+	}
+	tapdSource.BasicAuthEncoded, err = core.Encrypt(encKey, tapdSource.BasicAuthEncoded)
```
[the remark this answers is not in the document]
Comment (warren830) [reply plugins/tapd/api/tapd_sources.go:86]: fixed

**Review by klesh — APPROVED** [review]

**Pull request merged**